src/clients/kol.ts

```diff
@@ -87,95 +114,239 @@ function sanitiseBlueText(blueText: string | undefined): string {
+      if (await this.loggedIn()) return true;
+      if (this._isRollover) return false;
+      console.log(`Not logged in. Logging in as ${this._loginParameters.get("loginname")}`);
       try {
-        if (this._credentials.fetched < new Date().getTime() - 60000) {
-          const loginResponse = await axios("https://www.kingdomofloathing.com/login.php", {
-            method: "POST",
-            data: this._loginParameters,
-            maxRedirects: 0,
-            validateStatus: (status) => status === 302,
-          });
-          const sessionCookies = (loginResponse.headers["set-cookie"] || [])
-            .map((cookie: string) => cookie.split(";")[0])
-            .join("; ");
-          const apiResponse = await axios("https://www.kingdomofloathing.com/api.php", {
-            withCredentials: true,
-            headers: {
-              cookie: sessionCookies,
-            },
-            params: {
-              what: "status",
-              for: "OAF Discord bot for Kingdom of Loathing",
-            },
-          });
-          this._credentials = {
-            fetched: new Date().getTime(),
-            sessionCookies: sessionCookies,
-            pwdhash: apiResponse.data.pwd,
-          };
-        } else {
-          console.log("Blocked fetching new credentials");
-          console.log(
-            `${60000 + this._credentials.fetched - new Date().getTime()} milliseconds to new login`
-          );
-        }
-      } catch (error) {
-        console.log(error);
+        const loginResponse = await axios("https://www.kingdomofloathing.com/login.php", {
+          method: "POST",
+          data: this._loginParameters,
+          maxRedirects: 0,
+          validateStatus: (status) => status === 302,
+        });
+        const sessionCookies = (loginResponse.headers["set-cookie"] || [])
+          .map((cookie: string) => cookie.split(";")[0])
+          .join("; ");
+        const apiResponse = await axios("https://www.kingdomofloathing.com/api.php", {
+          withCredentials: true,
+          headers: {
+            cookie: sessionCookies,
+          },
+          params: {
+            what: "status",
+            for: `${this._loginParameters.get("loginname")} Chatbot`,
+          },
+        });
+        this._credentials = {
+          sessionCookies: sessionCookies,
+          pwdhash: apiResponse.data.pwd,
+        };
+        return true;
+      } catch {
+        console.log("Login failed. Checking if it's because of rollover.");
+        await this.rolloverCheck();
+        return false;
       }
     });
   }
 
-  private async makeCredentialedRequest(url: string, parameters: object) {
+  private chatBotStarted = false;
+
+  async startChatBot() {
+    if (this.chatBotStarted) return;
+    this.loopChatBot();
+    this.chatBotStarted = true;
+  }
+
+  private async loopChatBot() {
+    await Promise.all([this.checkWhispers(), this.checkKmails()]);
```

Comment:
Rather than checking kmails every 3 seconds, could we also check for the `[Event]` message that sends when you receive a kmail?

Reply:
We'd have to be checking for something every 3 secs, these pages are actually much lighter because they're just json